index/scorch/segment/zap/build.go

```diff
@@ -419,3 +426,104 @@ func persistDictionary(memSegment *mem.Segment, w *CountHashWriter, postingsLocs
 
 	return rv, nil
 }
+
+type docIDRange []uint64
+
+func (a docIDRange) Len() int           { return len(a) }
+func (a docIDRange) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a docIDRange) Less(i, j int) bool { return a[i] < a[j] }
+
+func persistDocValues(memSegment *mem.Segment, w *CountHashWriter,
+	chunkFactor uint32) (map[uint16]uint64, error) {
+	fieldChunkOffsets := make(map[uint16]uint64, len(memSegment.DocValueFields))
+	fdvEncoder := newChunkedContentCoder(uint64(chunkFactor), uint64(len(memSegment.Stored)-1))
+
+	for fieldID := range memSegment.DocValueFields {
+		field := memSegment.FieldsInv[fieldID]
+		docTermMap := make(map[uint64][]byte, 0)
+		dict, err := memSegment.Dictionary(field)
+		if err != nil {
+			return fieldChunkOffsets, err
+		}
+
+		dictItr := dict.Iterator()
+		next, err := dictItr.Next()
+		for err == nil && next != nil {
+			postings, err1 := dict.PostingsList(next.Term, nil)
+			if err1 != nil {
+				return fieldChunkOffsets, err
+			}
+
+			postingsItr := postings.Iterator()
+			nextPosting, err2 := postingsItr.Next()
+			for err2 == nil && nextPosting != nil {
+				docNum := nextPosting.Number()
+				docTermMap[docNum] = append(docTermMap[docNum], []byte(next.Term)...)
+				docTermMap[docNum] = append(docTermMap[docNum], termSeparator)
+				nextPosting, err2 = postingsItr.Next()
+			}
+			if err2 != nil {
+				return fieldChunkOffsets, err2
+			}
+
+			next, err = dictItr.Next()
+		}
+
+		if err != nil {
+			return fieldChunkOffsets, err
+		}
+		// sort wrt to docIDs
+		var docNumbers docIDRange
+		for k := range docTermMap {
+			docNumbers = append(docNumbers, k)
+		}
+		sort.Sort(docNumbers)
+
+		for _, docNum := range docNumbers {
+			err = fdvEncoder.Add(docNum, docTermMap[docNum])
+			if err != nil {
+				return fieldChunkOffsets, err
+			}
+		}
+
+		fieldChunkOffsets[fieldID] = uint64(w.Count())
+		fdvEncoder.Close()
+		// persist the doc value details for this field
+		_, err = fdvEncoder.Write(w)
+		if err != nil {
+			return fieldChunkOffsets, err
+		}
+		// resetting encoder for the next field
+		fdvEncoder.Reset()
+	}
+
+	return fieldChunkOffsets, nil
+}
+
+func persistFieldDocValues(w *CountHashWriter, chunkFactor uint32,
+	memSegment *mem.Segment) (uint64, error) {
+
+	fieldDvOffsets, err := persistDocValues(memSegment, w, chunkFactor)
+	if err != nil {
+		return math.MaxUint64, err
```

Comment:
any significance to this return value? if not i'd probably suggest just using 0 which is more common in Go when returning an error

Reply:
hm.. It was all zero's in my original code. For some reason I felt zero being a default value, it is better to specify an explicit value  through a different code., Hence initialised to "math.MaxUint64" .. But let me  change to '0' ..

Reply:
I don't think we ever check that value though in the case of an error, I think for people reading the code `0, err` makes more sense.  Think of it this way, everyone reading Go code has some idea what `return 0, err` means.  When I read `return math.MaxUint64, err` I immediately had to stop and think, I wonder why he's returning a non-zero value with the error.  But, when I looked closer, it's not used anywhere, so it seems like just using `0,err` reduces the amount of thinking for the person reading the code.